synchronizer/synchronizer.go

```diff
@@ -682,3 +730,113 @@ func (s *ClientSynchronizer) processVerifiedBatch(verifiedBatch etherman.Verifie
 		log.Fatalf("error storing the verifiedBatch in processVerifiedBatch. BlockNumber: %d, error: %s", verifiedBatch.BlockNumber, err.Error())
 	}
 }
+
+func (s *ClientSynchronizer) processTrustedBatch(trustedBatch *pb.GetBatchResponse, dbTx pgx.Tx) error {
+	log.Debugf("processing trusted batch: %v", trustedBatch.BatchNumber)
+	processCtx := state.ProcessingContext{
+		BatchNumber:    trustedBatch.BatchNumber,
+		Coinbase:       common.HexToAddress(trustedBatch.Sequencer),
+		Timestamp:      time.Unix(int64(trustedBatch.Timestamp), 0),
+		GlobalExitRoot: common.HexToHash(trustedBatch.GlobalExitRoot),
+	}
+	txs := []types.Transaction{}
+	for _, transaction := range trustedBatch.Transactions {
+		tx, err := state.DecodeTx(transaction.Encoded)
+		if err != nil {
+			return err
+		}
+		txs = append(txs, *tx)
+	}
+	trustedBatchL2Data, err := state.EncodeTransactions(txs)
+	if err != nil {
+		return err
+	}
+
+	batch, err := s.state.GetBatchByNumber(s.ctx, trustedBatch.BatchNumber, nil)
+	if err != nil && err != state.ErrStateNotSynchronized {
+		log.Warnf("failed to get batch %v from local trusted state. Error: %v", trustedBatch.BatchNumber, err)
+		return err
+	}
+
+	// check if batch needs to be synchronized
+	if batch != nil {
+		matchNumber := batch.BatchNumber == trustedBatch.BatchNumber
+		matchGER := batch.GlobalExitRoot.String() == trustedBatch.GlobalExitRoot
+		matchLER := batch.LocalExitRoot.String() == trustedBatch.LocalExitRoot
+		matchSR := batch.StateRoot.String() == trustedBatch.StateRoot
+		matchCoinbase := batch.Coinbase.String() == trustedBatch.Sequencer
+		matchTimestamp := uint64(batch.Timestamp.Unix()) == trustedBatch.Timestamp
+		matchL2Data := hex.EncodeToString(batch.BatchL2Data) == hex.EncodeToString(trustedBatchL2Data)
+
+		if matchNumber && matchGER && matchLER && matchSR &&
+			matchCoinbase && matchTimestamp && matchL2Data {
+			log.Debugf("batch %v already synchronized", trustedBatch.BatchNumber)
+			return nil
+		}
+		log.Infof("batch %v needs to be updated", trustedBatch.BatchNumber)
+	} else {
+		log.Infof("batch %v needs to be synchronized", trustedBatch.BatchNumber)
+	}
+
+	log.Debugf("resetting trusted state from batch %v", trustedBatch.BatchNumber)
+	if err := s.state.ResetTrustedState(s.ctx, trustedBatch.BatchNumber-1, dbTx); err != nil {
+		log.Errorf("failed to reset trusted state", trustedBatch.BatchNumber)
+		return err
+	}
```

Comment:
Why do you reset the trusted state here?? Maybe to remove the txs synced until this moment and proccess them again?? If it is, seems that it is not optimal. If not, I think we shouldn't reset the state

Reply:
As I've mentioned above, we need to discuss the `sync <==> state` interaction for this scenario. This PR uses a mix of the interface for sequencer and synchronizer. This is normal because when we designed the state interface we didn't have this scenario in mind.

Let's do a huddle when you're online

Reply:
When you say `it is not optimal` is It based on what? performance? maintenance? safeness? code readability? code complexity?

A previous trusted state synced from the trusted sequencer can also be affected by an L1 reorg or a trusted state reorg due to a forced batch. These scenarios would lead us to have processed `"wrong"/"invalid"` transactions from previously trusted state synchronization.

When the synchronization understands it's time to sync a batch from the trusted state, sure this is not the most performative solution, but it is the safest and simple thing to do until we have a good reason to change this behavior.

I recommend looking for two topics:
- Pareto Principle 80 20
- Early optimization is bad